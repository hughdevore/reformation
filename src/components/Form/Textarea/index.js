// @flow
import React, { Component } from 'react';
import { cx } from 'emotion';
import { textareaClass } from './styled';

type Props = {
  value: string,
  className?: string | null,
  onChange?: (value: string) => void,
  bindRef?: (element: any) => void,
};

type State = {
  value: string,
};

export default class Textarea extends Component<Props, State> {
  state = {
    value: this.props.value || '',
  };

  static getDerivedStateFromProps(nextProps, prevState) {
    if (nextProps.value === prevState.value) {
      return null;
    }
    return { value: nextProps.value || '' };
  }

  onChange = (e: { target: HTMLTextAreaElement }) => {
    const value = e.target.value || '';
    this.setState({ value }, () => {
      if (this.props.onChange) {
        this.props.onChange(value);
      }
    });
  };

  render() {
    const { className, onChange, value, bindRef, ...props } = this.props;
    return (
      <textarea
        {...props}
        ref={bindRef}
        className={cx(textareaClass, className)}
        onChange={this.onChange}
        value={this.state.value}
      />
    );
  }
}

Textarea.defaultProps = {
  className: null,
  onChange: () => null,
  bindRef: () => null,
};
