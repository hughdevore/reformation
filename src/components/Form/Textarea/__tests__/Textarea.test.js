import React from 'react';
import Textarea from 'components/Form/Textarea';
import { shallow } from 'enzyme';

const TEXT_VALUE = 'Run for the border.';

describe('Textarea', () => {
  test('empty', () => {
    const wrapper = shallow(<Textarea />);
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.state('value')).toEqual('');
  });

  test('add className', () => {
    const wrapper = shallow(<Textarea className="foo" />);
    expect(wrapper).toMatchSnapshot();
  });

  describe('onChange', () => {
    test('adding text', () => {
      const func = jest.fn();
      const wrapper = shallow(<Textarea onChange={func} />);
      const value = TEXT_VALUE;
      wrapper.simulate('change', { target: { value } });
      expect(wrapper.state('value')).toEqual(value);
    });

    test('removing text', () => {
      const func = jest.fn();
      const value = TEXT_VALUE;
      const wrapper = shallow(<Textarea onChange={func} value={value} />);
      wrapper.simulate('change', { target: { value: '' } });
      expect(wrapper.state('value')).toEqual('');
    });

    test('bad event', () => {
      const func = jest.fn();
      const value = TEXT_VALUE;
      const wrapper = shallow(<Textarea onChange={func} value={value} />);
      wrapper.simulate('change', { target: { foo: 'bar' } });
      expect(wrapper.state('value')).toEqual('');
    });

    test('bad type', () => {
      const func = jest.fn();
      const value = TEXT_VALUE;
      const wrapper = shallow(<Textarea onChange={func} value={value} />);
      wrapper.simulate('change', { target: { value: false } });
      expect(wrapper.state('value')).toEqual('');
    });
  });
});
