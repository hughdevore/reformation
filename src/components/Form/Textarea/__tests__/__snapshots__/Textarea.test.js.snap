// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Textarea add className 1`] = `
<textarea
  className="css-16crr32-inputStyles-textareaClass foo"
  onChange={[Function]}
  value=""
/>
`;

exports[`Textarea empty 1`] = `
<textarea
  className="css-16crr32-inputStyles-textareaClass"
  onChange={[Function]}
  value=""
/>
`;
